jsonrpc/server.go

```diff
@@ -187,10 +187,25 @@ func (s *Server) registerMethod(method Method) error {
 	return nil
 }
 
-// HandleReader processes a request to the server
+// Handle reads a JSON-RPC request from conn and writes the response.
+func (s *Server) Handle(ctx context.Context, conn io.ReadWriter) error {
+	resp, err := s.handle(ctx, conn)
+	if err != nil {
+		return err
+	}
+	if resp == nil {
+		return nil
+	}
+	if _, err = conn.Write(resp); err != nil {
+		return err
+	}
+	return nil
+}
+
+// handle processes a request to the server
 // It returns the response in a byte array, only returns an
 // error if it can not create the response byte array
-func (s *Server) HandleReader(ctx context.Context, reader io.Reader) ([]byte, error) {
+func (s *Server) handle(ctx context.Context, reader io.Reader) ([]byte, error) {
```

Comment:
How about this,

We keep the old API. `Handle(ctx context.Context, reader io.Reader)`
And still pass `reader` as connection to the handler. In the subscription handler, we try to "promote" it to a `io.ReadWriter` by casting it. (Like WS server tries to promote a http connection to a ws connection.) If cast fails,  reject the request it was probably sent with HTTP.

This way, we dont need to update `jsonrpc/http.go` at all